File: java/src/main/java/vsdk/toolkit/raycasting/stochasticRaytracing/StochasticRelaxation.java
package vsdk.toolkit.raycasting.stochasticRaytracing;

import vsdk.toolkit.common.ColorRgb;
import vsdk.toolkit.common.Error;
import vsdk.toolkit.tonemap.ToneMappingContext;

/**
Used for stochastic relaxation and for random walk radiosity
*/
public class StochasticRelaxation {
    public StochasticRaytracingMethod method; // Stochastic relaxation or random walks
    public WhatToShow show; // What to show and how to display the result
    public int inited; // Flag indicating whether initialised or not
    public int currentIteration;
    public ColorRgb unShotFlux;
    public ColorRgb totalFlux;
    public ColorRgb indirectImportanceWeightedUnShotFlux;
    public float unShotYmp;
    public float totalYmp; // Sum over all patches of area * importance
    public float sourceYmp;
    public int rayUnitsPerIt; // To increase or decrease initial nr of rays
    public int bidirectionalTransfers; // For bidirectional energy transfers
    public int constantControlVariate; // For constant control variate variance reduction
    public ColorRgb controlRadiance; // Constant control radiance value
    public int indirectOnly; // If to compute indirect illumination only
    public int weightedSampling; // If to do weighted sampling ala Powell and Swann / Spanier
    public int setSource; // For copying direct illumination to SOURCE_RAD(..) if computing only indirect illumination
    public Sampler4DSequence sequence; // Random number sequence
    public StochasticRaytracingApproximation approximationOrderType; // Radiosity approximation order
    public int importanceDriven; // If to use view-importance
    public int radianceDriven; // Radiance-driven importance propagation
    public int importanceUpdated; // Direct importance got updated or not?
    public int importanceUpdatedFromScratch; // Can either incremental or from scratch
    public int continuousRandomWalk; // Continuous or discrete random walk
    public RandomWalkEstimatorType randomWalkEstimatorType; // Shooting, gathering, gathering for free
    public RandomWalkEstimatorKind randomWalkEstimatorKind; // Collision, absorption, ...
    public int randomWalkNumLast; // For last-but-n and n-last RW estimators
    public int discardIncremental; // First iteration (incremental steps) results are discarded in later computations
    public int incrementalUsesImportance; // View-importance is used already for the first iteration (incremental steps). This may confuse the merging heuristic or other things
    public int naiveMerging; // Results of different iterations are merged solely based on the number of rays shot
    public long initialNumberOfRays; // For first iteration step
    public long raysPerIteration; // For later iterations
    public long importanceRaysPerIteration; // For later iterations for importance
    public long tracedRays; // Total number of traced rays
    public long prevTracedRays; // Previous total of traced rays
    public long importanceTracedRays;
    public long prevImportanceTracedRays;
    public long tracedPaths; // Number of traced random walks in random walk rad.
    public long numberOfMisses; // Rays disappearing to background
    public int doNonDiffuseFirstShot; // Initial shooting pass handles non-diffuse lights
    public int initialLightSourceSamples; // Initial shot samples per light source
    public long lastClock; // For computation timings (nanoseconds)
    public float cpuSeconds; // CPU time spent in calculations
    public ToneMappingContext toneMapOptions;

    public StochasticRelaxation() {
        method = null;
        show = null;
        inited = 0;
        currentIteration = 0;
        unShotFlux = new ColorRgb();
        totalFlux = new ColorRgb();
        indirectImportanceWeightedUnShotFlux = new ColorRgb();
        unShotYmp = 0.0f;
        totalYmp = 0.0f;
        sourceYmp = 0.0f;
        rayUnitsPerIt = 0;
        bidirectionalTransfers = 0;
        constantControlVariate = 0;
        controlRadiance = new ColorRgb();
        indirectOnly = 0;
        weightedSampling = 0;
        setSource = 0;
        sequence = null;
        approximationOrderType = null;
        importanceDriven = 0;
        radianceDriven = 0;
        importanceUpdated = 0;
        importanceUpdatedFromScratch = 0;
        continuousRandomWalk = 0;
        randomWalkEstimatorType = null;
        randomWalkEstimatorKind = null;
        randomWalkNumLast = 0;
        discardIncremental = 0;
        incrementalUsesImportance = 0;
        naiveMerging = 0;
        initialNumberOfRays = 0L;
        raysPerIteration = 0L;
        importanceRaysPerIteration = 0L;
        tracedRays = 0L;
        prevTracedRays = 0L;
        importanceTracedRays = 0L;
        prevImportanceTracedRays = 0L;
        tracedPaths = 0L;
        numberOfMisses = 0L;
        doNonDiffuseFirstShot = 0;
        initialLightSourceSamples = 0;
        lastClock = 0L;
        cpuSeconds = 0.0f;
        toneMapOptions = null;
    }

    public static void setActiveState(StochasticRelaxation state) {
        activeStatePtr = state;
    }

    public static StochasticRelaxation activeState() {
        if ( activeStatePtr == null ) {
            Error.fatal(-1, "StochasticRelaxation::activeState", "Stochastic relaxation state was not initialized");
        }
        return activeStatePtr;
    }

    private static StochasticRelaxation activeStatePtr = null;
}

File: typescript/base/src/main/vsdk/toolkit/raycasting/stochasticRaytracing/StochasticRelaxation.ts
import { ColorRgb } from "../../common/color/ColorRgb";
import { Logger as VsdkLogger } from "../../common/logging/Logger";
import { ToneMappingContext } from "../../tonemap/ToneMappingContext";
import { RandomWalkEstimatorKind } from "./RandomWalkEstimatorKind";
import { RandomWalkEstimatorType } from "./RandomWalkEstimatorType";
import { Sampler4DSequence } from "./Sampler4DSequence";
import { StochasticRaytracingApproximation } from "./StochasticRaytracingApproximation";
import { StochasticRaytracingMethod } from "./StochasticRaytracingMethod";
import { WhatToShow } from "./WhatToShow";

/**
Used for stochastic relaxation and for random walk radiosity
*/
export class StochasticRelaxation {
  public method: StochasticRaytracingMethod | null;
  public show: WhatToShow | null;
  public inited: number;
  public currentIteration: number;
  public unShotFlux: ColorRgb;
  public totalFlux: ColorRgb;
  public indirectImportanceWeightedUnShotFlux: ColorRgb;
  public unShotYmp: number;
  public totalYmp: number;
  public sourceYmp: number;
  public rayUnitsPerIt: number;
  public bidirectionalTransfers: number;
  public constantControlVariate: number;
  public controlRadiance: ColorRgb;
  public indirectOnly: number;
  public weightedSampling: number;
  public setSource: number;
  public sequence: Sampler4DSequence | null;
  public approximationOrderType: StochasticRaytracingApproximation | null;
  public importanceDriven: number;
  public radianceDriven: number;
  public importanceUpdated: number;
  public importanceUpdatedFromScratch: number;
  public continuousRandomWalk: number;
  public randomWalkEstimatorType: RandomWalkEstimatorType | null;
  public randomWalkEstimatorKind: RandomWalkEstimatorKind | null;
  public randomWalkNumLast: number;
  public discardIncremental: number;
  public incrementalUsesImportance: number;
  public naiveMerging: number;
  public initialNumberOfRays: number;
  public raysPerIteration: number;
  public importanceRaysPerIteration: number;
  public tracedRays: number;
  public prevTracedRays: number;
  public importanceTracedRays: number;
  public prevImportanceTracedRays: number;
  public tracedPaths: number;
  public numberOfMisses: number;
  public doNonDiffuseFirstShot: number;
  public initialLightSourceSamples: number;
  public lastClock: number;
  public cpuSeconds: number;
  public toneMapOptions: ToneMappingContext | null;

  public constructor() {
    this.method = null;
    this.show = null;
    this.inited = 0;
    this.currentIteration = 0;
    this.unShotFlux = new ColorRgb();
    this.totalFlux = new ColorRgb();
    this.indirectImportanceWeightedUnShotFlux = new ColorRgb();
    this.unShotYmp = 0.0;
    this.totalYmp = 0.0;
    this.sourceYmp = 0.0;
    this.rayUnitsPerIt = 0;
    this.bidirectionalTransfers = 0;
    this.constantControlVariate = 0;
    this.controlRadiance = new ColorRgb();
    this.indirectOnly = 0;
    this.weightedSampling = 0;
    this.setSource = 0;
    this.sequence = null;
    this.approximationOrderType = null;
    this.importanceDriven = 0;
    this.radianceDriven = 0;
    this.importanceUpdated = 0;
    this.importanceUpdatedFromScratch = 0;
    this.continuousRandomWalk = 0;
    this.randomWalkEstimatorType = null;
    this.randomWalkEstimatorKind = null;
    this.randomWalkNumLast = 0;
    this.discardIncremental = 0;
    this.incrementalUsesImportance = 0;
    this.naiveMerging = 0;
    this.initialNumberOfRays = 0;
    this.raysPerIteration = 0;
    this.importanceRaysPerIteration = 0;
    this.tracedRays = 0;
    this.prevTracedRays = 0;
    this.importanceTracedRays = 0;
    this.prevImportanceTracedRays = 0;
    this.tracedPaths = 0;
    this.numberOfMisses = 0;
    this.doNonDiffuseFirstShot = 0;
    this.initialLightSourceSamples = 0;
    this.lastClock = 0;
    this.cpuSeconds = 0.0;
    this.toneMapOptions = null;
  }

  public static setActiveState(state: StochasticRelaxation): void {
    StochasticRelaxation.activeStatePtr = state;
  }

  public static activeState(): StochasticRelaxation {
    if (StochasticRelaxation.activeStatePtr === null) {
      VsdkLogger.fatal(-1, "StochasticRelaxation::activeState", "Stochastic relaxation state was not initialized");
    }
    return StochasticRelaxation.activeStatePtr!;
  }

  private static activeStatePtr: StochasticRelaxation | null = null;
}
